### PR TITLE
Different defend plays

### LINKED ISSUE #1103: Defend Shot

### LINKED ISSUE #1104: Defend pass

### PR DESCRIPTION
**Pull request by jordiluong**
Split Defend play into DefendPass and DefendShot plays with different behaviour.

Closes #1103 
Closes #1104

### CODE REVIEW

File: src/stp/new_plays/DefendPass.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+namespace rtt::ai::stp::play {
+
+DefendPass::DefendPass() : Play() {
+    startPlayInvariants.clear();
+    startPlayInvariants.emplace_back(std::make_unique<invariant::NormalPlayGameStateInvariant>());
+    startPlayInvariants.emplace_back(std::make_unique<invariant::BallCloseToThemInvariant>());
+
+    keepPlayInvariants.clear();
+    keepPlayInvariants.emplace_back(std::make_unique<invariant::NormalPlayGameStateInvariant>());
```
Comment (timovdk) [inline src/stp/new_plays/DefendPass.cpp:23]: I think this means that we will only switch plays when we are not in normal_play

Comment (jordiluong) [reply src/stp/new_plays/DefendPass.cpp:23]: Yes, you are right, the same for DefendShot. We somehow need a way to determine when we are done defending hmmm...

Comment (Yuhanun) [reply src/stp/new_plays/DefendPass.cpp:23]: Unrelated but damn we should figure something out for these disgusting emplace_back's with make_unique all over the place, it just looks _really_ bad.

File: src/stp/new_plays/DefendPass.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+    roles = std::array<std::unique_ptr<Role>, stp::control_constants::MAX_ROBOT_COUNT>{std::make_unique<role::Keeper>(role::Keeper("keeper")),
```
Comment (Yuhanun) [inline src/stp/new_plays/DefendPass.cpp:26]: I'm confident you dont have to explicitly mark the type here.

File: src/stp/new_plays/DefendPass.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+Dealer::FlagMap DefendPass::decideRoleFlags() const noexcept {
+    Dealer::FlagMap flagMap;
+    Dealer::DealerFlag keeperFlag(DealerFlagTitle::KEEPER, DealerFlagPriority::KEEPER);
+    Dealer::DealerFlag closeToBallFlag(DealerFlagTitle::CLOSE_TO_BALL, DealerFlagPriority::HIGH_PRIORITY);
+    Dealer::DealerFlag closeToOurGoalFlag(DealerFlagTitle::CLOSE_TO_OUR_GOAL, DealerFlagPriority::HIGH_PRIORITY);
+    Dealer::DealerFlag closeToTheirGoalFlag(DealerFlagTitle::CLOSE_TO_THEIR_GOAL, DealerFlagPriority::LOW_PRIORITY);
+    Dealer::DealerFlag not_important(DealerFlagTitle::ROBOT_TYPE_50W, DealerFlagPriority::LOW_PRIORITY);
+
+    flagMap.insert({"keeper", {keeperFlag}});
```
Comment (Yuhanun) [inline src/stp/new_plays/DefendPass.cpp:49]: flagMap["keeper"] = keeperFlag;

File: src/stp/new_plays/DefendPass.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+    stpInfos["defender_2"].setPositionToDefend(field.getOurBottomGoalSide());
+    stpInfos["defender_2"].setEnemyRobot(enemyClosestToBall);
+    stpInfos["defender_2"].setBlockDistance(HALFWAY);
+
+    // When the ball moves, one defender tries to intercept the ball
+    for (auto &role : roles) {
+        auto roleName = role->getName();
+        if (roleName.find("defender") != std::string::npos) {
+            // TODO: Improve choice of intercept robot based on trajectory and intercept position
```
Comment (Yuhanun) [inline src/stp/new_plays/DefendPass.cpp:87]:
```cpp
if (x == y) {
    continue;
}
// other stuff
```

**Review by Yuhanun — APPROVED** [review]
lgtm.

**Review by timovdk — APPROVED** [review]

**Pull request merged**